## Notebook02c

### Setup

Run all of the following before starting the notebook.

In [1]:
! wget -q -nc https://raw.githubusercontent.com/taylor-arnold/fds-py/refs/heads/main/funs.py

In [2]:
import numpy as np
import polars as pl

from funs import *
from plotnine import *
from polars import col as c
theme_set(theme_minimal())

ub = "https://raw.githubusercontent.com/taylor-arnold/fds-py-nb/refs/heads/main/"

In [3]:
film = pl.read_csv(ub + "data/criterion.csv").filter(~c.rating_imdb.is_null())
film

shape: (1_470, 18)
┌────────────┬────────────┬──────┬────────────┬───┬────────────┬───────────┬───────────┬───────────┐
│ imdb_id    ┆ title      ┆ year ┆ language   ┆ … ┆ wikipedia_ ┆ wikipedia ┆ budget_ra ┆ box_offic │
│ ---        ┆ ---        ┆ ---  ┆ ---        ┆   ┆ descriptio ┆ _extract  ┆ w         ┆ e_raw     │
│ str        ┆ str        ┆ i64  ┆ str        ┆   ┆ n          ┆ ---       ┆ ---       ┆ ---       │
│            ┆            ┆      ┆            ┆   ┆ ---        ┆ str       ┆ i64       ┆ i64       │
│            ┆            ┆      ┆            ┆   ┆ str        ┆           ┆           ┆           │
╞════════════╪════════════╪══════╪════════════╪═══╪════════════╪═══════════╪═══════════╪═══════════╡
│ tt0012349  ┆ The Kid    ┆ 1921 ┆ None|Engli ┆ … ┆ 1921       ┆ The Kid   ┆ 250000    ┆ null      │
│            ┆            ┆      ┆ sh         ┆   ┆ silent     ┆ is a 1921 ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ film by    ┆ American  ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ Charlie    ┆ sil…      ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ Ch…        ┆           ┆           ┆           │
│ tt0012364  ┆ The        ┆ 1921 ┆ None|Swedi ┆ … ┆ 1921 film  ┆ The       ┆ null      ┆ null      │
│            ┆ Phantom    ┆      ┆ sh         ┆   ┆ by Victor  ┆ Phantom   ┆           ┆           │
│            ┆ Carriage   ┆      ┆            ┆   ┆ Sjöström   ┆ Carriage  ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆            ┆ is a      ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆            ┆ 1921…     ┆           ┆           │
│ tt0013257  ┆ Häxan      ┆ 1922 ┆ Swedish|Da ┆ … ┆ Swedish    ┆ Häxan is  ┆ 2000000   ┆ null      │
│            ┆            ┆      ┆ nish       ┆   ┆ 1922       ┆ a 1922    ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ silent     ┆ Swedish-D ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ horror     ┆ anish…    ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ ess…       ┆           ┆           ┆           │
│ tt0014429  ┆ Safety     ┆ 1923 ┆ English    ┆ … ┆ 1923       ┆ Safety    ┆ 121000    ┆ null      │
│            ┆ Last!      ┆      ┆            ┆   ┆ American   ┆ Last! is  ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ silent     ┆ a 1923    ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ romantic … ┆ America…  ┆           ┆           │
│ tt0014624  ┆ A Woman of ┆ 1923 ┆ English    ┆ … ┆ 1923 drama ┆ A Woman   ┆ 351000    ┆ 634000    │
│            ┆ Paris: A   ┆      ┆            ┆   ┆ film by    ┆ of Paris  ┆           ┆           │
│            ┆ Drama of   ┆      ┆            ┆   ┆ Charlie    ┆ is a 1923 ┆           ┆           │
│            ┆ F…         ┆      ┆            ┆   ┆ Cha…       ┆ sil…      ┆           ┆           │
│ …          ┆ …          ┆ …    ┆ …          ┆ … ┆ …          ┆ …         ┆ …         ┆ …         │
│ tt27958252 ┆ Caught by  ┆ 2024 ┆ Mandarin|E ┆ … ┆ 2024 film  ┆ Caught by ┆ null      ┆ null      │
│            ┆ the Tides  ┆      ┆ nglish     ┆   ┆ by Jia     ┆ the Tides ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆ Zhangke    ┆ is a 2024 ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆            ┆ …         ┆           ┆           │
│ tt28607951 ┆ Anora      ┆ 2024 ┆ English|Ru ┆ … ┆ 2024 film  ┆ Anora is  ┆ null      ┆ null      │
│            ┆            ┆      ┆ ssian|Arme ┆   ┆ by Sean    ┆ a 2024    ┆           ┆           │
│            ┆            ┆      ┆ nian       ┆   ┆ Baker      ┆ American  ┆           ┆           │
│            ┆            ┆      ┆            ┆   ┆            ┆ roman…    ┆           ┆           │
│ tt28618488 ┆ Vermiglio  ┆ 2024 ┆ Italian

### Questions

In this notebook, we will practice data manipulation using a dataset of films from the Criterion Collection. The `film` dataset contains information about each movie including its title, director, year of release, runtime, and IMDB rating.

Start by sorting the films by their IMDB rating, with the lowest-rated films at the top.

In [5]:
film_rating = (
    film
    .sort(c.rating_imdb)
)
film_rating

shape: (1_470, 18)
┌───────────┬────────────┬──────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐
│ imdb_id   ┆ title      ┆ year ┆ language   ┆ … ┆ wikipedia_ ┆ wikipedia_ ┆ budget_ra ┆ box_offic │
│ ---       ┆ ---        ┆ ---  ┆ ---        ┆   ┆ descriptio ┆ extract    ┆ w         ┆ e_raw     │
│ str       ┆ str        ┆ i64  ┆ str        ┆   ┆ n          ┆ ---        ┆ ---       ┆ ---       │
│           ┆            ┆      ┆            ┆   ┆ ---        ┆ str        ┆ i64       ┆ i64       │
│           ┆            ┆      ┆            ┆   ┆ str        ┆            ┆           ┆           │
╞═══════════╪════════════╪══════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡
│ tt0062492 ┆ Wild 90    ┆ 1968 ┆ English    ┆ … ┆ 1968 film  ┆ Wild 90 is ┆ null      ┆ null      │
│           ┆            ┆      ┆            ┆   ┆ by Norman  ┆ a 1968 exp ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Mailer     ┆ erimental… ┆           ┆           │
│ tt0064373 ┆ All        ┆ 1969 ┆ Japanese   ┆ … ┆ 1969 film  ┆ All        ┆ null      ┆ null      │
│           ┆ Monsters   ┆      ┆            ┆   ┆ by Ishirō  ┆ Monsters   ┆           ┆           │
│           ┆ Attack     ┆      ┆            ┆   ┆ Honda      ┆ Attack  is ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ a 1969…    ┆           ┆           │
│ tt0063195 ┆ Genocide   ┆ 1968 ┆ Japanese   ┆ … ┆ 1968       ┆ Genocide   ┆ null      ┆ null      │
│           ┆            ┆      ┆            ┆   ┆ Japanese   ┆ is a 1968  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ Japanese   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ s…         ┆           ┆           │
│ tt0064625 ┆ Maidstone  ┆ 1970 ┆ English    ┆ … ┆ 1970 film  ┆ Maidstone  ┆ null      ┆ null      │
│           ┆            ┆      ┆            ┆   ┆ by Norman  ┆ is a 1970  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Mailer     ┆ American   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ i…         ┆           ┆           │
│ tt2006181 ┆ Jellyfish  ┆ 2013 ┆ Japanese   ┆ … ┆ 2013       ┆ Jellyfish  ┆ null      ┆ null      │
│           ┆ Eyes       ┆      ┆            ┆   ┆ Japanese   ┆ Eyes  is a ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ 2013 Japa… ┆           ┆           │
│ …         ┆ …          ┆ …    ┆ …          ┆ … ┆ …          ┆ …          ┆ …         ┆ …         │
│ tt0048434 ┆ Night and  ┆ 1956 ┆ French     ┆ … ┆ 1956       ┆ Night and  ┆ null      ┆ null      │
│           ┆ Fog        ┆      ┆            ┆   ┆ French     ┆ Fog is a   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ 1956       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ French…    ┆           ┆           │
│ tt0056058 ┆ Harakiri   ┆ 1962 ┆ Japanese   ┆ … ┆ Japanese   ┆ Harakiri   ┆ null      ┆ 15000     │
│           ┆            ┆      ┆            ┆   ┆ jidaigeki  ┆ is a 1962  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ Japanese   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ j…         ┆           ┆           │
│ tt0102926 ┆ The        ┆ 1991 ┆ English|La ┆ … ┆ 1991       ┆ The        ┆ 19000000  ┆ 130742922 │
│           ┆ Silence of ┆      ┆ tin        ┆   ┆ horror     ┆ Silence of ┆           ┆           │
│           ┆ the Lambs  ┆      ┆            ┆   ┆ film by    ┆ the Lambs  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Jonathan   ┆ is a …     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ D…         ┆            ┆           ┆           │
│ tt0090015 ┆ Shoah      ┆ 1985 ┆ German|H

Now, sort the films by runtime with the longest films at the top.

In [7]:
film_runtime = (
    film
    .sort(c.runtime_raw, descending = True)
)
film_runtime

shape: (1_470, 18)
┌───────────┬────────────┬──────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐
│ imdb_id   ┆ title      ┆ year ┆ language   ┆ … ┆ wikipedia_ ┆ wikipedia_ ┆ budget_ra ┆ box_offic │
│ ---       ┆ ---        ┆ ---  ┆ ---        ┆   ┆ descriptio ┆ extract    ┆ w         ┆ e_raw     │
│ str       ┆ str        ┆ i64  ┆ str        ┆   ┆ n          ┆ ---        ┆ ---       ┆ ---       │
│           ┆            ┆      ┆            ┆   ┆ ---        ┆ str        ┆ i64       ┆ i64       │
│           ┆            ┆      ┆            ┆   ┆ str        ┆            ┆           ┆           │
╞═══════════╪════════════╪══════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡
│ tt0023261 ┆ No Blood   ┆ 1932 ┆ None|Japan ┆ … ┆ 1932       ┆ No Blood   ┆ null      ┆ null      │
│           ┆ Relation   ┆      ┆ ese        ┆   ┆ Japanese   ┆ Relation   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ is a 1932  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ J…         ┆           ┆           │
│ tt0024214 ┆ Apart from ┆ 1933 ┆ None|Japan ┆ … ┆ 1933       ┆ Apart From ┆ null      ┆ null      │
│           ┆ You        ┆      ┆ ese        ┆   ┆ Japanese   ┆ You  is a  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ 1933 Japa… ┆           ┆           │
│ tt0025338 ┆ Street     ┆ 1934 ┆ None       ┆ … ┆ 1934       ┆ Street     ┆ null      ┆ null      │
│           ┆ Without    ┆      ┆            ┆   ┆ Japanese   ┆ Without    ┆           ┆           │
│           ┆ End        ┆      ┆            ┆   ┆ film       ┆ End  is a  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ 1934 …     ┆           ┆           │
│ tt0029002 ┆ History Is ┆ 1937 ┆ English|It ┆ … ┆ 1937 film  ┆ History Is ┆ null      ┆ null      │
│           ┆ Made at    ┆      ┆ alian|Fren ┆   ┆ by Frank   ┆ Made at    ┆           ┆           │
│           ┆ Night      ┆      ┆ ch         ┆   ┆ Borzage    ┆ Night is a ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ …          ┆           ┆           │
│ tt0031828 ┆ The Tunnel ┆ 1940 ┆ English    ┆ … ┆ 1940 film  ┆ The Proud  ┆ null      ┆ null      │
│           ┆            ┆      ┆            ┆   ┆ by Pen     ┆ Valley is  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Tennyson   ┆ a 1940     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ Eal…       ┆           ┆           │
│ …         ┆ …          ┆ …    ┆ …          ┆ … ┆ …          ┆ …          ┆ …         ┆ …         │
│ tt0056662 ┆ Vive le    ┆ 1962 ┆ French     ┆ … ┆ 1962       ┆ Vive le    ┆ null      ┆ null      │
│           ┆ tour       ┆      ┆            ┆   ┆ French     ┆ Tour is a  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ 1962       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ French …   ┆           ┆           │
│ tt0075122 ┆ The        ┆ 1976 ┆ Persian    ┆ … ┆ 1976       ┆ The        ┆ null      ┆ null      │
│           ┆ Colours    ┆      ┆            ┆   ┆ Iranian    ┆ Colours is ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ a 1976     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ Iranian …  ┆           ┆           │
│ tt0063551 ┆ Saute ma   ┆ 1971 ┆ French     ┆ … ┆ Belgian    ┆ Chantal    ┆ null      ┆ null      │
│           ┆ ville      ┆      ┆            ┆   ┆ film       ┆ Anne       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ director,  ┆ Akerman    ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ screenw…   ┆ was a Bel… ┆           ┆           │
│ tt0068354 ┆ La chambre ┆ 1972 ┆ French  

Select only the films that are less than 60 minutes long (i.e., short films).

In [9]:
short_films = (
    film
    .filter(c.runtime_raw < 60)
)
short_films

shape: (26, 18)
┌───────────┬────────────┬──────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐
│ imdb_id   ┆ title      ┆ year ┆ language   ┆ … ┆ wikipedia_ ┆ wikipedia_ ┆ budget_ra ┆ box_offic │
│ ---       ┆ ---        ┆ ---  ┆ ---        ┆   ┆ descriptio ┆ extract    ┆ w         ┆ e_raw     │
│ str       ┆ str        ┆ i64  ┆ str        ┆   ┆ n          ┆ ---        ┆ ---       ┆ ---       │
│           ┆            ┆      ┆            ┆   ┆ ---        ┆ str        ┆ i64       ┆ i64       │
│           ┆            ┆      ┆            ┆   ┆ str        ┆            ┆           ┆           │
╞═══════════╪════════════╪══════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡
│ tt0021576 ┆ À Propos   ┆ 1930 ┆ None       ┆ … ┆ 1930 film  ┆ À propos   ┆ null      ┆ null      │
│           ┆ de Nice    ┆      ┆            ┆   ┆            ┆ de Nice is ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ a 1930     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ sil…       ┆           ┆           │
│ tt0022036 ┆ Flunky,    ┆ 1931 ┆ None|Japan ┆ … ┆ 1931       ┆ Flunky,    ┆ null      ┆ null      │
│           ┆ Work Hard! ┆      ┆ ese        ┆   ┆ Japanese   ┆ Work Hard! ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ is a 1931  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ …          ┆           ┆           │
│ tt0024803 ┆ Zero for   ┆ 1933 ┆ French     ┆ … ┆ 1933       ┆ Zero for   ┆ null      ┆ null      │
│           ┆ Conduct    ┆      ┆            ┆   ┆ French     ┆ Conduct is ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ a 1933     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ Fre…       ┆           ┆           │
│ tt0028445 ┆ A Day in   ┆ 1936 ┆ French     ┆ … ┆ 1946       ┆ Partie de  ┆ null      ┆ null      │
│           ┆ the        ┆      ┆            ┆   ┆ French     ┆ campagne   ┆           ┆           │
│           ┆ Country    ┆      ┆            ┆   ┆ film       ┆ is a       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ French…    ┆           ┆           │
│ tt0038182 ┆ The Men    ┆ 1945 ┆ Japanese   ┆ … ┆ 1945       ┆ The Men    ┆ null      ┆ null      │
│           ┆ Who Tread  ┆      ┆            ┆   ┆ Japanese   ┆ Who Tread  ┆           ┆           │
│           ┆ on the     ┆      ┆            ┆   ┆ film by    ┆ on the     ┆           ┆           │
│           ┆ Tiger…     ┆      ┆            ┆   ┆ Akira Ku…  ┆ Tiger…     ┆           ┆           │
│ …         ┆ …          ┆ …    ┆ …          ┆ … ┆ …          ┆ …          ┆ …         ┆ …         │
│ tt0079704 ┆ Paul       ┆ 1979 ┆ English|Ma ┆ … ┆ 1979 film  ┆ Paul       ┆ null      ┆ null      │
│           ┆ Robeson:   ┆      ┆ ndarin     ┆   ┆            ┆ Robeson:   ┆           ┆           │
│           ┆ Tribute to ┆      ┆            ┆   ┆            ┆ Tribute to ┆           ┆           │
│           ┆ an Ar…     ┆      ┆            ┆   ┆            ┆ an Ar…     ┆           ┆           │
│ tt0081204 ┆ Mur murs   ┆ 1981 ┆ French|Eng ┆ … ┆ 1980 docum ┆ Mur Murs   ┆ null      ┆ null      │
│           ┆            ┆      ┆ lish|Itali ┆   ┆ entary     ┆ is a 1981  ┆           ┆           │
│           ┆            ┆      ┆ an|Spanish ┆   ┆ film by    ┆ documentar ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Agnès…     ┆ y…         ┆           ┆           │
│ tt0083727 ┆ Chambre    ┆ 1982 ┆ English|Fr ┆ … ┆ 1982       ┆ Room 666   ┆ null      ┆ null      │
│           ┆ 666        ┆      ┆ ench|Germa ┆   ┆ French     ┆ is a 1982  ┆           ┆           │
│           ┆            ┆      ┆ n|Italian| ┆   ┆ film       ┆ documentar ┆           ┆           │
│           ┆            ┆      ┆ …          

Select all films directed by Charles Chaplin.

In [11]:
chaplin_films = (
    film
    .filter(c.director == "Charles Chaplin")
)
chaplin_films

shape: (8, 18)
┌───────────┬────────────┬──────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐
│ imdb_id   ┆ title      ┆ year ┆ language   ┆ … ┆ wikipedia_ ┆ wikipedia_ ┆ budget_ra ┆ box_offic │
│ ---       ┆ ---        ┆ ---  ┆ ---        ┆   ┆ descriptio ┆ extract    ┆ w         ┆ e_raw     │
│ str       ┆ str        ┆ i64  ┆ str        ┆   ┆ n          ┆ ---        ┆ ---       ┆ ---       │
│           ┆            ┆      ┆            ┆   ┆ ---        ┆ str        ┆ i64       ┆ i64       │
│           ┆            ┆      ┆            ┆   ┆ str        ┆            ┆           ┆           │
╞═══════════╪════════════╪══════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡
│ tt0012349 ┆ The Kid    ┆ 1921 ┆ None|Engli ┆ … ┆ 1921       ┆ The Kid is ┆ 250000    ┆ null      │
│           ┆            ┆      ┆ sh         ┆   ┆ silent     ┆ a 1921     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film by    ┆ American   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Charlie    ┆ sil…       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Ch…        ┆            ┆           ┆           │
│ tt0014624 ┆ A Woman of ┆ 1923 ┆ English    ┆ … ┆ 1923 drama ┆ A Woman of ┆ 351000    ┆ 634000    │
│           ┆ Paris: A   ┆      ┆            ┆   ┆ film by    ┆ Paris is a ┆           ┆           │
│           ┆ Drama of   ┆      ┆            ┆   ┆ Charlie    ┆ 1923 sil…  ┆           ┆           │
│           ┆ F…         ┆      ┆            ┆   ┆ Cha…       ┆            ┆           ┆           │
│ tt0018773 ┆ The Circus ┆ 1928 ┆ None|Engli ┆ … ┆ 1928 film  ┆ The Circus ┆ null      ┆ null      │
│           ┆            ┆      ┆ sh         ┆   ┆ by Charlie ┆ is a 1928  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Chaplin    ┆ silent ro… ┆           ┆           │
│ tt0021749 ┆ City       ┆ 1931 ┆ None|Engli ┆ … ┆ 1931       ┆ City       ┆ 1500000   ┆ 5000000   │
│           ┆ Lights     ┆      ┆ sh         ┆   ┆ American   ┆ Lights is  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ silent     ┆ a 1931     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ American…  ┆           ┆           │
│ tt0027977 ┆ Modern     ┆ 1936 ┆ English    ┆ … ┆ 1936       ┆ Modern     ┆ 1500000   ┆ 1400000   │
│           ┆ Times      ┆      ┆            ┆   ┆ comedy     ┆ Times is a ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film by    ┆ 1936       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Charlie    ┆ America…   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Ch…        ┆            ┆           ┆           │
│ tt0032553 ┆ The Great  ┆ 1940 ┆ English|Es ┆ … ┆ 1940       ┆ The Great  ┆ 2000000   ┆ 5000000   │
│           ┆ Dictator   ┆      ┆ peranto|La ┆   ┆ American   ┆ Dictator   ┆           ┆           │
│           ┆            ┆      ┆ tin        ┆   ┆ film by    ┆ is a 1940  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Charlie …  ┆ A…         ┆           ┆           │
│ tt0039631 ┆ Monsieur   ┆ 1947 ┆ English|Fr ┆ … ┆ 1947 film  ┆ Monsieur   ┆ 2000000   ┆ null      │
│           ┆ Verdoux    ┆      ┆ ench|Latin ┆   ┆ by Charlie ┆ Verdoux is ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Chaplin    ┆ a 1947     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆            ┆ Ame…       ┆           ┆           │
│ tt0044837 ┆ Limelight  ┆ 1952 ┆ English    ┆ … ┆ 1952 film  ┆ Limelight  ┆ 900000    ┆ 1000000   │
│           ┆            ┆      ┆            ┆   ┆ by Charlie ┆ is a 1952  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Chaplin    ┆ American   ┆           ┆           │
│           ┆            ┆      ┆            ┆

Select all films directed by either Charles Chaplin or Jean-Luc Godard.

In [17]:
filter_films = (
    film
    .filter((c.director == "Charles Chaplin") | (c.director == "Jean-Luc Godard"))
)
filter_films

shape: (19, 18)
┌───────────┬────────────┬──────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐
│ imdb_id   ┆ title      ┆ year ┆ language   ┆ … ┆ wikipedia_ ┆ wikipedia_ ┆ budget_ra ┆ box_offic │
│ ---       ┆ ---        ┆ ---  ┆ ---        ┆   ┆ descriptio ┆ extract    ┆ w         ┆ e_raw     │
│ str       ┆ str        ┆ i64  ┆ str        ┆   ┆ n          ┆ ---        ┆ ---       ┆ ---       │
│           ┆            ┆      ┆            ┆   ┆ ---        ┆ str        ┆ i64       ┆ i64       │
│           ┆            ┆      ┆            ┆   ┆ str        ┆            ┆           ┆           │
╞═══════════╪════════════╪══════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡
│ tt0012349 ┆ The Kid    ┆ 1921 ┆ None|Engli ┆ … ┆ 1921       ┆ The Kid is ┆ 250000    ┆ null      │
│           ┆            ┆      ┆ sh         ┆   ┆ silent     ┆ a 1921     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film by    ┆ American   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Charlie    ┆ sil…       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Ch…        ┆            ┆           ┆           │
│ tt0014624 ┆ A Woman of ┆ 1923 ┆ English    ┆ … ┆ 1923 drama ┆ A Woman of ┆ 351000    ┆ 634000    │
│           ┆ Paris: A   ┆      ┆            ┆   ┆ film by    ┆ Paris is a ┆           ┆           │
│           ┆ Drama of   ┆      ┆            ┆   ┆ Charlie    ┆ 1923 sil…  ┆           ┆           │
│           ┆ F…         ┆      ┆            ┆   ┆ Cha…       ┆            ┆           ┆           │
│ tt0018773 ┆ The Circus ┆ 1928 ┆ None|Engli ┆ … ┆ 1928 film  ┆ The Circus ┆ null      ┆ null      │
│           ┆            ┆      ┆ sh         ┆   ┆ by Charlie ┆ is a 1928  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Chaplin    ┆ silent ro… ┆           ┆           │
│ tt0021749 ┆ City       ┆ 1931 ┆ None|Engli ┆ … ┆ 1931       ┆ City       ┆ 1500000   ┆ 5000000   │
│           ┆ Lights     ┆      ┆ sh         ┆   ┆ American   ┆ Lights is  ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ silent     ┆ a 1931     ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film       ┆ American…  ┆           ┆           │
│ tt0027977 ┆ Modern     ┆ 1936 ┆ English    ┆ … ┆ 1936       ┆ Modern     ┆ 1500000   ┆ 1400000   │
│           ┆ Times      ┆      ┆            ┆   ┆ comedy     ┆ Times is a ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ film by    ┆ 1936       ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Charlie    ┆ America…   ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Ch…        ┆            ┆           ┆           │
│ …         ┆ …          ┆ …    ┆ …          ┆ … ┆ …          ┆ …          ┆ …         ┆ …         │
│ tt0060675 ┆ Masculine  ┆ 1966 ┆ French|Swe ┆ … ┆ 1966 film  ┆ Masculin   ┆ null      ┆ null      │
│           ┆ Feminine   ┆      ┆ dish|Engli ┆   ┆ by         ┆ féminin:   ┆           ┆           │
│           ┆            ┆      ┆ sh         ┆   ┆ Jean-Luc   ┆ 15         ┆           ┆           │
│           ┆            ┆      ┆            ┆   ┆ Godard     ┆ Specific … ┆           ┆           │
│ tt0060304 ┆ 2 or 3     ┆ 1967 ┆ French|Ita ┆ … ┆ 1967 film  ┆ Two or     ┆ null      ┆ null      │
│           ┆ Things I   ┆      ┆ lian|Engli ┆   ┆ by         ┆ Three      ┆           ┆           │
│           ┆ Know About ┆      ┆ sh         ┆   ┆ Jean-Luc   ┆ Things I   ┆           ┆           │
│           ┆ Her        ┆      ┆            ┆   ┆ Godard     ┆ Know Abo…  ┆           ┆           │
│ tt0062480 ┆ Weekend    ┆ 1967 ┆ French     ┆ … ┆ Film by    ┆ Weekend is ┆ null      ┆ null      │
│           ┆            ┆      ┆            ┆   ┆ Jean-Luc   ┆ a 1967     ┆           ┆           │
│           ┆            ┆      ┆            

Find the highest-rated film from each year. Sort the final result by year so it is easier to read.

In [23]:
highest_rated = (
    film
    .sort(c.rating_imdb)
    .group_by(c.year)
    .head(n=1)
)
highest_rated

shape: (102, 18)
┌──────┬───────────┬────────────┬────────────┬───┬────────────┬────────────┬───────────┬───────────┐
│ year ┆ imdb_id   ┆ title      ┆ language   ┆ … ┆ wikipedia_ ┆ wikipedia_ ┆ budget_ra ┆ box_offic │
│ ---  ┆ ---       ┆ ---        ┆ ---        ┆   ┆ descriptio ┆ extract    ┆ w         ┆ e_raw     │
│ i64  ┆ str       ┆ str        ┆ str        ┆   ┆ n          ┆ ---        ┆ ---       ┆ ---       │
│      ┆           ┆            ┆            ┆   ┆ ---        ┆ str        ┆ i64       ┆ i64       │
│      ┆           ┆            ┆            ┆   ┆ str        ┆            ┆           ┆           │
╞══════╪═══════════╪════════════╪════════════╪═══╪════════════╪════════════╪═══════════╪═══════════╡
│ 1996 ┆ tt0115964 ┆ Crash      ┆ English|Sw ┆ … ┆ Canadian   ┆ Crash is a ┆ null      ┆ null      │
│      ┆           ┆            ┆ edish      ┆   ┆ erotic     ┆ 1996       ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆ thriller   ┆ Canadian   ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆ film …     ┆ eroti…     ┆           ┆           │
│ 2001 ┆ tt0243255 ┆ Fat Girl   ┆ French|Ita ┆ … ┆ 2001 film  ┆ Fat Girl   ┆ null      ┆ null      │
│      ┆           ┆            ┆ lian|Engli ┆   ┆ by         ┆ is a 2001  ┆           ┆           │
│      ┆           ┆            ┆ sh         ┆   ┆ Catherine  ┆ coming-of- ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆ Breilla…   ┆ a…         ┆           ┆           │
│ 1981 ┆ tt0082050 ┆ Orderly or ┆ Persian    ┆ … ┆ 1981       ┆ Orderly or ┆ null      ┆ null      │
│      ┆           ┆ Disorderly ┆            ┆   ┆ Iranian    ┆ Disorderly ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆ film       ┆ is a 198…  ┆           ┆           │
│ 1993 ┆ tt0107162 ┆ Once Upon  ┆ Cantonese  ┆ … ┆ 1993 Hong  ┆ Once Upon  ┆ null      ┆ null      │
│      ┆           ┆ a Time in  ┆            ┆   ┆ Kong film  ┆ a Time in  ┆           ┆           │
│      ┆           ┆ China IV   ┆            ┆   ┆ by Yuen    ┆ China IV   ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆ Bu…        ┆ i…         ┆           ┆           │
│ 1948 ┆ tt0311384 ┆ Kalpana    ┆ Hindi      ┆ … ┆ 1948       ┆ Kalpana    ┆ null      ┆ null      │
│      ┆           ┆            ┆            ┆   ┆ Indian     ┆ (transl. I ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆ film       ┆ magination ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆            ┆ ) …        ┆           ┆           │
│ …    ┆ …         ┆ …          ┆ …          ┆ … ┆ …          ┆ …          ┆ …         ┆ …         │
│ 1937 ┆ tt0029064 ┆ Dark Sands ┆ English    ┆ … ┆ 1937       ┆ Jericho is ┆ null      ┆ null      │
│      ┆           ┆            ┆            ┆   ┆ British    ┆ a 1937     ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆ film       ┆ British    ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆            ┆ dram…      ┆           ┆           │
│ 2021 ┆ tt6910282 ┆ Bergman    ┆ English|Sw ┆ … ┆ 2021 film  ┆ Bergman    ┆ null      ┆ null      │
│      ┆           ┆ Island     ┆ edish|Fren ┆   ┆ by Mia Han ┆ Island is  ┆           ┆           │
│      ┆           ┆            ┆ ch         ┆   ┆ sen-Løve   ┆ a 2021     ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆            ┆ roman…     ┆           ┆           │
│ 2012 ┆ tt1843287 ┆ Like       ┆ Japanese   ┆ … ┆ 2012       ┆ Like       ┆ 4800000   ┆ 239056    │
│      ┆           ┆ Someone in ┆            ┆   ┆ Japanese   ┆ Someone in ┆           ┆           │
│      ┆           ┆ Love       ┆            ┆   ┆ film       ┆ Love is a  ┆           ┆           │
│      ┆           ┆            ┆            ┆   ┆            ┆ Fren…      ┆           ┆           │
│ 1947 ┆ tt0039834 ┆ A Ship to  ┆ Swedish   

For each director, compute the average IMDB rating of their films and the number of films they have in the collection.

In [26]:
rat_by_dir = (
    film
    .group_by(c.director)
    .agg(
        num_films = c.title.count(),
        avg_rating = c.rating_imdb.mean()
    )
)
rat_by_dir

shape: (650, 3)
┌─────────────────────────────┬───────────┬────────────┐
│ director                    ┆ num_films ┆ avg_rating │
│ ---                         ┆ ---       ┆ ---        │
│ str                         ┆ u32       ┆ f64        │
╞═════════════════════════════╪═══════════╪════════════╡
│ Martha Coolidge             ┆ 1         ┆ 8.1        │
│ Wei Lo|Chia-Hsiang Wu       ┆ 1         ┆ 6.9        │
│ Albert Lamorisse            ┆ 3         ┆ 6.766667   │
│ Kimiyoshi Yasuda            ┆ 5         ┆ 7.22       │
│ Edmund Goulding             ┆ 1         ┆ 7.7        │
│ …                           ┆ …         ┆ …          │
│ Hajime Satô                 ┆ 1         ┆ 6.1        │
│ JR|Agnès Varda              ┆ 1         ┆ 7.8        │
│ Fred C. Newmeyer|Sam Taylor ┆ 2         ┆ 7.8        │
│ Jûzô Itami                  ┆ 2         ┆ 7.55       │
│ Darius Marder               ┆ 1         ┆ 7.7        │
└─────────────────────────────┴───────────┴────────────┘

We often want to group data by decade rather than individual years. To compute the decade from a year, we can use the expression `c.year // 10 * 10`. The `//` operator performs integer division (dividing and dropping the remainder), so for example `1987 // 10` gives `198`, and multiplying by 10 gives `1980`. Create a new column containing the decade each film was released in.

In [29]:
film_decades = (
    film
    .with_columns(decade = c.year // 10 * 10)
)
film_decades

shape: (1_470, 19)
┌────────────┬─────────────┬──────┬────────────┬───┬────────────┬────────────┬────────────┬────────┐
│ imdb_id    ┆ title       ┆ year ┆ language   ┆ … ┆ wikipedia_ ┆ budget_raw ┆ box_office ┆ decade │
│ ---        ┆ ---         ┆ ---  ┆ ---        ┆   ┆ extract    ┆ ---        ┆ _raw       ┆ ---    │
│ str        ┆ str         ┆ i64  ┆ str        ┆   ┆ ---        ┆ i64        ┆ ---        ┆ i64    │
│            ┆             ┆      ┆            ┆   ┆ str        ┆            ┆ i64        ┆        │
╞════════════╪═════════════╪══════╪════════════╪═══╪════════════╪════════════╪════════════╪════════╡
│ tt0012349  ┆ The Kid     ┆ 1921 ┆ None|Engli ┆ … ┆ The Kid is ┆ 250000     ┆ null       ┆ 1920   │
│            ┆             ┆      ┆ sh         ┆   ┆ a 1921     ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ American   ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ sil…       ┆            ┆            ┆        │
│ tt0012364  ┆ The Phantom ┆ 1921 ┆ None|Swedi ┆ … ┆ The        ┆ null       ┆ null       ┆ 1920   │
│            ┆ Carriage    ┆      ┆ sh         ┆   ┆ Phantom    ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ Carriage   ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ is a 1921… ┆            ┆            ┆        │
│ tt0013257  ┆ Häxan       ┆ 1922 ┆ Swedish|Da ┆ … ┆ Häxan is a ┆ 2000000    ┆ null       ┆ 1920   │
│            ┆             ┆      ┆ nish       ┆   ┆ 1922 Swedi ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ sh-Danish… ┆            ┆            ┆        │
│ tt0014429  ┆ Safety      ┆ 1923 ┆ English    ┆ … ┆ Safety     ┆ 121000     ┆ null       ┆ 1920   │
│            ┆ Last!       ┆      ┆            ┆   ┆ Last! is a ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ 1923       ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ America…   ┆            ┆            ┆        │
│ tt0014624  ┆ A Woman of  ┆ 1923 ┆ English    ┆ … ┆ A Woman of ┆ 351000     ┆ 634000     ┆ 1920   │
│            ┆ Paris: A    ┆      ┆            ┆   ┆ Paris is a ┆            ┆            ┆        │
│            ┆ Drama of F… ┆      ┆            ┆   ┆ 1923 sil…  ┆            ┆            ┆        │
│ …          ┆ …           ┆ …    ┆ …          ┆ … ┆ …          ┆ …          ┆ …          ┆ …      │
│ tt27958252 ┆ Caught by   ┆ 2024 ┆ Mandarin|E ┆ … ┆ Caught by  ┆ null       ┆ null       ┆ 2020   │
│            ┆ the Tides   ┆      ┆ nglish     ┆   ┆ the Tides  ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ is a 2024  ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ …          ┆            ┆            ┆        │
│ tt28607951 ┆ Anora       ┆ 2024 ┆ English|Ru ┆ … ┆ Anora is a ┆ null       ┆ null       ┆ 2020   │
│            ┆             ┆      ┆ ssian|Arme ┆   ┆ 2024       ┆            ┆            ┆        │
│            ┆             ┆      ┆ nian       ┆   ┆ American   ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ roman…     ┆            ┆            ┆        │
│ tt28618488 ┆ Vermiglio   ┆ 2024 ┆ Italian|La ┆ … ┆ Vermiglio  ┆ null       ┆ null       ┆ 2020   │
│            ┆             ┆      ┆ tin        ┆   ┆ , also     ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ known as   ┆            ┆            ┆        │
│            ┆             ┆      ┆            ┆   ┆ The …      ┆            ┆            ┆        │
│ tt32085997 ┆ Misericordi ┆ 2024 ┆ French     ┆ … ┆ Misericord ┆ null       ┆ null       ┆ 2020   │
│            ┆ a           ┆      ┆            ┆   ┆ ia is a    ┆            ┆            ┆        │
│            ┆             ┆      ┆       

Now, count how many films in the collection were released in each decade.

In [32]:
film_decade_count = (
    film_decades
    .group_by(c.decade)
    .agg(
        decade_count = c.title.count()
    )
)
film_decade_count

shape: (11, 2)
┌────────┬──────────────┐
│ decade ┆ decade_count │
│ ---    ┆ ---          │
│ i64    ┆ u32          │
╞════════╪══════════════╡
│ 1980   ┆ 171          │
│ 1930   ┆ 119          │
│ 2010   ┆ 55           │
│ 2020   ┆ 47           │
│ 1960   ┆ 303          │
│ …      ┆ …            │
│ 1940   ┆ 122          │
│ 1920   ┆ 19           │
│ 2000   ┆ 86           │
│ 1970   ┆ 232          │
│ 1950   ┆ 162          │
└────────┴──────────────┘

Find the directors with more than 5 films in the collection, showing their average IMDB rating. Sort the results so the highest-rated directors appear at the top.

In [33]:
avg_dir_rat = (
    film
    .group_by(c.director)
    .agg(
        avg_rating = c.rating_imdb.mean(),
        count = pl.len()
    )
    .filter(c.count > 5)
    .sort(c.avg_rating, descending = True)
)
avg_dir_rat

shape: (48, 3)
┌─────────────────────┬────────────┬───────┐
│ director            ┆ avg_rating ┆ count │
│ ---                 ┆ ---        ┆ ---   │
│ str                 ┆ f64        ┆ u32   │
╞═════════════════════╪════════════╪═══════╡
│ Charles Chaplin     ┆ 8.05       ┆ 8     │
│ Satyajit Ray        ┆ 7.981818   ┆ 11    │
│ Kenji Mizoguchi     ┆ 7.777778   ┆ 9     │
│ Richard Linklater   ┆ 7.766667   ┆ 6     │
│ Carl Theodor Dreyer ┆ 7.683333   ┆ 6     │
│ …                   ┆ …          ┆ …     │
│ Pier Paolo Pasolini ┆ 6.941667   ┆ 12    │
│ Nagisa Ôshima       ┆ 6.85       ┆ 8     │
│ Keisuke Kinoshita   ┆ 6.842857   ┆ 7     │
│ David Cronenberg    ┆ 6.742857   ┆ 7     │
│ Chantal Akerman     ┆ 6.7375     ┆ 8     │
└─────────────────────┴────────────┴───────┘

For each director with more than 5 films in the collection, show their highest-rated film. Hint: If you sort by rating before grouping, the first title in each group will be that director's highest-rated film.

In [39]:
highest_film_dir = (
    film
    .sort(c.rating_imdb, descending = True)
    .group_by(c.director)
    .agg(
        count = pl.len(),
        highest_title = c.title.first()
    )
    .filter(c.count > 5)
)
highest_film_dir

shape: (48, 3)
┌────────────────────┬───────┬───────────────────────┐
│ director           ┆ count ┆ highest_title         │
│ ---                ┆ ---   ┆ ---                   │
│ str                ┆ u32   ┆ str                   │
╞════════════════════╪═══════╪═══════════════════════╡
│ Richard Linklater  ┆ 6     ┆ Before Sunrise        │
│ Wim Wenders        ┆ 12    ┆ Paris, Texas          │
│ Seijun Suzuki      ┆ 7     ┆ Youth of the Beast    │
│ Jim Jarmusch       ┆ 6     ┆ Down by Law           │
│ Keisuke Kinoshita  ┆ 7     ┆ Twenty-Four Eyes      │
│ …                  ┆ …     ┆ …                     │
│ Koreyoshi Kurahara ┆ 6     ┆ Intimidation          │
│ Alfred Hitchcock   ┆ 7     ┆ Rebecca               │
│ Akira Kurosawa     ┆ 24    ┆ Seven Samurai         │
│ Louis Malle        ┆ 13    ┆ Au Revoir les Enfants │
│ Jean Renoir        ┆ 10    ┆ The Rules of the Game │
└────────────────────┴───────┴───────────────────────┘

By default, only the first few rows of a result are displayed. To see more rows, we can pipe the result through the `print_rows` function. Modify your previous answer to display all of the results.

In [41]:
highest_film_dir = (
    film
    .sort(c.rating_imdb, descending = True)
    .group_by(c.director)
    .agg(
        count = pl.len(),
        highest_title = c.title.first()
    )
    .filter(c.count > 5)
    .sort(c.count, descending = True)
    .pipe(print_rows)
)
highest_film_dir

shape: (48, 3)
┌─────────────────────────────────┬───────┬─────────────────────────────────┐
│ director                        ┆ count ┆ highest_title                   │
│ ---                             ┆ ---   ┆ ---                             │
│ str                             ┆ u32   ┆ str                             │
╞═════════════════════════════════╪═══════╪═════════════════════════════════╡
│ Ingmar Bergman                  ┆ 39    ┆ Scenes from a Marriage          │
│ Abbas Kiarostami                ┆ 26    ┆ Close-Up                        │
│ Akira Kurosawa                  ┆ 24    ┆ Seven Samurai                   │
│ Yasujirô Ozu                    ┆ 20    ┆ Late Spring                     │
│ Agnès Varda                     ┆ 17    ┆ The Beaches of Agnès            │
│ Louis Malle                     ┆ 13    ┆ Au Revoir les Enfants           │
│ Pier Paolo Pasolini             ┆ 12    ┆ Accattone                       │
│ Wim Wenders                     ┆ 12    ┆ Paris, Texas                    │
│ Satyajit Ray                    ┆ 11    ┆ The World of Apu                │
│ Jean-Luc Godard                 ┆ 11    ┆ Vivre sa vie                    │
│ Jean Renoir                     ┆ 10    ┆ The Rules of the Game           │
│ Wes Anderson                    ┆ 10    ┆ The Grand Budapest Hotel        │
│ Rainer Werner Fassbinder        ┆ 10    ┆ Ali: Fear Eats the Soul         │
│ Kenji Misumi                    ┆ 10    ┆ Lone Wolf and Cub: Baby Cart a… │
│ Aki Kaurismäki                  ┆ 9     ┆ The Bohemian Life               │
│ Kenji Mizoguchi                 ┆ 9     ┆ Sansho the Bailiff              │
│ Ernst Lubitsch                  ┆ 8     ┆ To Be or Not to Be              │
│ Roberto Rossellini              ┆ 8     ┆ Rome, Open City                 │
│ Chantal Akerman                 ┆ 8     ┆ Jeanne Dielman, 23, quai du Co… │
│ Charles Chaplin                 ┆ 8     ┆ City Lights                     │
│ Nagisa Ôshima                   ┆ 8     ┆ Death by Hanging                │
│ Martin Scorsese                 ┆ 8     ┆ The Last Waltz                  │
│ Wong Kar-Wai                    ┆ 7     ┆ In the Mood for Love            │
│ Michael Powell|Emeric Pressbur… ┆ 7     ┆ The Red Shoes                   │
│ Samuel Fuller                   ┆ 7     ┆ Pickup on South Street          │
│ Seijun Suzuki                   ┆ 7     ┆ Youth of the Beast              │
│ Michelangelo Antonioni          ┆ 7     ┆ La Notte                        │
│ Alfred Hitchcock                ┆ 7     ┆ Rebecca                         │
│ Federico Fellini                ┆ 7     ┆ La Strada                       │
│ David Lean                      ┆ 7     ┆ Brief Encounter                 │
│ Masaki Kobayashi                ┆ 7     ┆ Harakiri                        │
│ David Cronenberg                ┆ 7     ┆ A History of Violence           │
│ François Truffaut               ┆ 7     ┆ The 400 Blows                   │
│ John Cassavetes                 ┆ 7     ┆ A Woman Under the Influence     │
│ Keisuke Kinoshita               ┆ 7     ┆ Twenty-Four Eyes                │
│ David Lynch                     ┆ 6     ┆ Mulholland Drive                │
│ Jacques Demy                    ┆ 6     ┆ The Umbrellas of Cherbourg      │
│ Koreyoshi Kurahara              ┆ 6     ┆ Intimidation                    │
│ Richard Linklater               ┆ 6     ┆ Before Sunrise                  │
│ Michael Haneke                  ┆ 6     ┆ The Seventh Continent           │
│ Robert Bresson                  ┆ 6     ┆ A Man Escaped                   │
│ Carl Theodor Dreyer             ┆ 6     ┆ Ordet                           │
│ Julien Duvivier                 ┆ 6     ┆ Panique                         │
│ Jim Jarmusch                    ┆ 6     ┆ Down by Law                     │
│ Vittorio De Sica                ┆ 6     ┆ Bicycle Thieves                 │
│ Marlon Riggs                    ┆ 6     ┆ Ethnic Notions                  │
│ Mikio